### ISSUE TITLE
Captcha on gettoken

### ISSUE THREAD

**Issue by Taloth** [rarbg/torrentapi#11]
I've been getting reports from user that gettoken errors out with a captcha message.
Is there anything sonarr could do to avoid this?

I haven't seen it myself yet, so i don't know if there's a relation with a ratelimit or something similar.

**Comment by rarbg** [comment]
Unless you give me ip ranges of the users i cant tell you why cloudflare is returning captchas for them

**Comment by Taloth** [comment]
66.55.134.* is one I grabbed from irc, no guarantee it's the one running sonarr. I'll check our forum later.

**Comment by rarbg** [comment]
Thats a :
A) VPS from vultr
B) Dedicated server from choopa
C) Random VPN that was abused and thats why Cloudflare blocked it with a CAPTCHA
Blocks are completely automated for A+B+C so there is nothing i can do
You can tell them to switch IP regions if its C , if its A+B ask them why they were abusing the API ( by abusing i mean 1k+ req / hour for example )

**Comment by Taloth** [comment]
Tnx, that's useful.

Sonarr doesn't store cookies by default, so could be a shared ip thing. Without the cookie, cloudflare's anti-ddos won't be able to differentiate individual clients.

We're gathering some more ips from users and whether they use vpn.

**Comment by a92nj2efnfsa** [comment]
The IP address that Taloth gave is correct, that is my IP address. I'm using private internet access and I'm not abusing the IP. I've disconnected and reconnected several times and gotten different IP addresses all with the same result. IPs I've been assigned by them that have not worked:

66.55.134.*
64.237.37.*
108.61.68.*

**Comment by rarbg** [comment]
@a92nj2efnfsa thousands of users are using those ips , they are not used only by you and someone abused them and thats why you get a captcha! 
For the past 48 hours there are 291490 requests from those ranges.

**Comment by rarbg** [comment]
Oh and some more stats @a92nj2efnfsa @Taloth 
Out of the 36 million requests for the past 48 hours - 5.1 million were made from AS20473 or Choopa, LLC ( incl the 3 ranges that i was given ) + multiple other ranges in US , FR , UK , NL , DE , JP , AU 
The previous 24 hours before those 48 - only a mere 10k req / day were coming from that network
Only 486 users(ips) out of 50k daily unique got the captcha which is below 0.95%
Im sorry but there is no way i can compromise the performance of the API and manually remove the captcha on that network. 
I will monitor it closely and keep you updated

**Comment by a92nj2efnfsa** [comment]
I wonder why the requests suddenly skyrocketed. That's really weird. I 100% understand not wanting to compromise the performance of the API - do you have any suggestions for me so that I can get it working again for myself?

**Comment by Taloth** [comment]
I expected an increase after kat went down, but 19mil req/day seems quite excessive, that's a 35% increase. I'd imagine a lot of kat users were using VPNs, and that may account for some of the increase but not all of it...

Do you have any stats per app_id? And the get_token api? Basically to see how many are legitimate clients.

I briefly checked the internet for some options, but it appears that the cloudflare clearance cookie is tied to the Useragent as well, so I can't just add a cookie field to the settings.
All I can think of right now is make Sonarr log a human readable error.

**Comment by rarbg** [comment]
@Taloth 
Can you give me your email or just send an email to support@rarbg.com where we can discuss it ? There are some things i have to say that are not for a public issue

**Comment by FredBro** [comment]
FYI: I'm using a VPN as well at 107.191.36.* and I'm using MCM. I've been getting "RARBG API is not responding or returning errors!"

I've written the developer and opened a bug report here on Github for MediaCenterMaster but the developer pointed me to this thread.

It started about the time kat went down as well. I know of two other MCM users who are experiencing this problem on the MCM forums as well

**Comment by rarbg** [comment]
@FredBro you have to read carefully what i said. 107.191.36. is still part of Choopa LLC.
Just connect to another geographical node in your PIA client - for example NL

**Comment by a92nj2efnfsa** [comment]
@FredBro @rarbg what I ended up doing was just googling "free http proxy" and I added that in sonarr. That allowed me to pull down the rss feed and still use PIA.

**Comment by Taloth** [comment]
I really wouldn't use free public proxies.

**Comment by FredBro** [comment]
There's also no place to set a proxy in MCM. It would have to be at the OS level. And I agree, there are a lot of anti free proxy pages on the Internet telling of horror stories about malware.

Fred




On Mon, Aug 8, 2016 at 7:46 AM -0400, "Taloth" <notifications@github.com<mailto:notifications@github.com>> wrote:


I really wouldn't use free public proxies.

-

**Comment by a92nj2efnfsa** [comment]
@Taloth  just out of curiosity why? (Not saying you're wrong, just wanted to know your thoughts.) my reasoning for not caring was it's just pulling down the rss feed and it's doing it over my PIA VPN anyway.

**Comment by Taloth** [comment]
It's not just the rarbg api calls. You're proxying everything, including tvdb updates, xem and our own services. And potentially other indexers as well

**Comment by a92nj2efnfsa** [comment]
@Taloth yea, good point, but if @rarbg is blocking those PIA IPs not sure what any other option is at this point...

**Comment by Quaternions-FTW** [comment]
I was getting the same error using sonarr because I am connected to the a PIA VPN server which is connected with Choopa, LLC. I solved this by going to 

Settings>General>Proxy Settings
and inputting the following


**Use Proxy >** YES
**Proxy Type >** Socks5 (This option supports Tor)
**Hostname >** proxy-nl.privateinternetaccess.com
**Port >** 1080
**Username >** x#######
**Password >** ################
**Addresses for the proxy to ignore >**
**Bypass Proxy for Local Addresses >** YES

I left "Addresses for proxy to ignore" blank, but you could add thetvdb.com or others if you wish for sonarr to not use the proxy.  

If using PIA, you should use the "SOCKS5" username that is found on your PIA profile page. It will be the same as your normal PIA username, except it will begin with "x" instead of "p". Also the "SOCKS5" username will have a different password than your normal password. This can also be found in your PIA profile page.

**Comment by a92nj2efnfsa** [comment]
@Quaternions-FTW awesome, you just had to go and let @rarbg know that. Now they are blocking that as well.

**Comment by Quaternions-FTW** [comment]
Not sure if it is related or coincidental, but in my sonarr logs, I am now receiving this html output:


# Error 502 Ray ID: ################ • 2016-08-10 03:47:18 UTC
## Bad gateway
You
### Browser
Working
Amsterdam
### CloudFlare
Working
torrentapi.org
### Host
Error
## What happened?

The web server reported a bad gateway error.
## What can I do?

Please try again in a few minutes.

CloudFlare Ray ID: ################ • Your IP: 109.201.138.227 • Performance & security by CloudFlare

**Comment by odaysawaqed** [comment]
Getting this error as well
Error 502 Ray ID: 2d010c732b8a473a • 2016-08-10 05:29:18 UTC
Bad gateway

I tried to make a request from both my private ip and my VPN SP's ip
Both returned the same error!

According to the logs, this started happening on 8/9/16 at 5pm EST time

**Comment by Taloth** [comment]
If one of you would be so kind to test Sonarr develop version 2.​0.​0.​4284, via PIA of course. You'll have to go to the settings page first, as shown in the PR I linked above.

Please note that the cookie expires after a week and needs to be refreshed, once we confirmed everything is working @rarbg will change that to 30 days.
There will be no way around that, you'll have to manually refresh it monthly.

**Comment by odaysawaqed** [comment]
Seems a little weird to me, i had recently changed my sonarr to the develop branch and it seems to work just fine. I looked at my version and its at 2.0.0.4280 currently.  The version 2.​0.​0.​4284 was probably released just now.  I don't think the PIA is an issue here nor it has anything related to the issue you you guys are working on. I only mentioned it earlier because i was trying to give as much info to others having the same issue. Didn't realize the topic was in regards to a different issue. 
I was directed here by Google. My fault.

I will test it and report back once finished.

**Comment by Taloth** [comment]
Your issue had nothing to do with what we were discussing, you were just running into some temporary gateway timeouts, back then the api was just down for a while.
This issue is exclusively for ppl who's VPN IPs got blocked by Cloudflare captcha challenges.

I appreciate your willingness to test, but even if your test comes back positive, I still need someone else to verify it for the captcha.

**Comment by odaysawaqed** [comment]
Using IPA VPN on version 2.​0.​0.​4284, I followed your PR I linked above RARBG seems to be added just fine. Hitting the test button returns "Testing RARBG succeeded" but Sonarr doesn't seem to notice the Captcha Response, click Refresh Token doesn't seem to give any feedback.



Here is the log: 
[sonarr.debug.txt](https://github.com/rarbg/torrentapi/files/416786/sonarr.debug.txt)
[sonarr.trace.txt](https://github.com/rarbg/torrentapi/files/416787/sonarr.trace.txt)

**Comment by Taloth** [comment]
There is no captcha response, coz there is no captcha challenge. The "Testing Rarbg succeeded" means it's already fine without a captcha token.

**Comment by odaysawaqed** [comment]
So is there anything i can do to help here? can i force the captcha challenge ?

**Comment by Taloth** [comment]
No you can't, thank you. It has to be tested by users that were affected by the original issue.

**Comment by a92nj2efnfsa** [comment]
@Taloth what do I type in the branch text field to get the develop build? Also, how do I get back to master? is it easy to get back to the master branch?

**Comment by Taloth** [comment]
If you're using the build-in updater, then just `develop`.  See [here](https://github.com/Sonarr/Sonarr/wiki/Release-Branches) for more details.

You can't switch back immediately, due to potential database migrations. But once you got .4284 installed, you can change the field back to `master`, then you'll effectively switch back once the next master version gets released.

**Comment by a92nj2efnfsa** [comment]
@Taloth do you want me to do 4284 or 4286?

**Comment by Taloth** [comment]
Just the latest, so 4286. Sry, should've said that in my previous post.

**Comment by a92nj2efnfsa** [comment]
@Taloth never mind, I just noticed my only option was to install the lasted dev build anyway. I installed 4286. I went to the rarbg indexer and hit refresh and no token shows up in the text field. I saw nothing in the logs so then I changed the verbose level to debug and I see this:

16-8-15 16:53:25.0|Debug|Api|[POST] /api/indexer/action/checkCaptcha: 200.OK (345 ms)
16-8-15 16:53:25.8|Debug|Api|[POST] /api/indexer/action/checkCaptcha: 200.OK (358 ms)
16-8-15 16:53:25.9|Debug|Api|[POST] /api/indexer/action/checkCaptcha: 200.OK (330 ms)
16-8-15 16:53:26.2|Debug|Api|[POST] /api/indexer/action/checkCaptcha: 200.OK (358 ms)
16-8-15 16:53:26.5|Debug|Api|[POST] /api/indexer/action/checkCaptcha: 200.OK (391 ms)
16-8-15 16:53:27.0|Debug|Api|[POST] /api/indexer/action/checkCaptcha: 200.OK (423 ms)

(I hit the button a couple of times to see if I could find it in the log) - however, I did find out that PIA allows you access to their SOCKS5 proxy. So, I was using their SOCKS5 proxy setup and I was able to get the rarbg stuff just fine that way. So, I thought maybe this is why it was not getting a new token. 

I then decided to stop the openvpn service and restart it giving me a new IP on PIA. But for whatever reason I am unable to replicate the issue now. It just keeps succeeding and not failing. I'll keep playing with it tonight and see if I can replicate the issue and then I will try the captcha thing you built in.

But, regardless of all that, thank you for trying to fix this issue. I really appreciate it. I saw you stayed up all night coding it and I wanted you to know that's awesome and appreciated.

**Comment by a92nj2efnfsa** [comment]
@Taloth I want to clarify from my previous post that I was thinking that since I was able to get the token when the PIA SOCKS5 information was filled in that maybe it was somehow cached after I disabled the SOCKS5 proxy, but I guess that makes no sense as my IP would have changed anyway... I currently have the proxy disabled and will keep it that way. Once the issue comes back, I can then continue to test the captcha token.

**Comment by Taloth** [comment]
Thank you for trying to test it. Seems like the PIA IP addresses you're getting atm aren't blocked by CloudFlare. Please let me know once it errors out again.
The token I have expires tomorrow night, then I'll know whether Sonarr gives the proper errors.

I'll see about making it more clear that a token isn't always required. To prevent ppl spamming the Refresh button (which in turn spams torrentapi, so I don't want that).

_PS: Not to make you feel guilty or anything, only a peek behind the curtain: It actually took over half a dozen evenings, my guess is I spend around 40h on it total. It was worth it, the code I wrote will allow us to implement a selection UI for newznab categories more easily in the future. Not that I expect that to be implemented anytime soon, but it's nice to have the infrastructure ready._ :sweat_smile:

**Comment by rarbg** [comment]
@Taloth great work! If you want i can force captcha for any user so they can test ? It will be easier i think ?

**Comment by a92nj2efnfsa** [comment]
@Taloth ok, understood, I'll keep bouncing the openVPN service until I can get an IP that's blocked so I can effectively test it with you. I'll report back tomorrow with the results.

That's a crazy amount of time! Thanks again and good to hear about it helping with future builds :)

@rarbg can you block 208.167.254.108

**Comment by Taloth** [comment]
Tnx but no, the reason I want it tested naturally is because cloudflare has different threat levels and may throw up different defences on soft threats, unlike the hard 'i want this ip have a captcha'.

So it's a final sanity check before I declare it fixed and go on holiday :)

**Comment by Taloth** [comment]
@rarbg Can you extend it to 30 days? It neatly expires and logs errors, as intended. Gonna tweak the UI a bit but it should be ok as is.

**Comment by rarbg** [comment]
Done @Taloth

**Comment by mikesalo23** [comment]
Hi @rarbg - we're getting the captcha from 24.192.54.*

Any idea if you could check on it for me?

**Comment by Taloth** [comment]
Btw. @rarbg These captcha fixes have been release to master now. Feel free to remove the test ip from the blacklist.

I'll also close this issue.
